### PR TITLE
Add GreetingGenerator

### PR DESCRIPTION
**Pull request by gbirke**
The class generates greetings based on lastname, salutation and title.
This is needed for all mails that are sent to users.

### CODE REVIEW

**Comment by wmde-manicki** [comment]
I don't know if this is relevant but it might be worth keeping in mind that there might be different types of greeting text. See for instance: https://github.com/wmde/fundraising/issues/776 where at least in some cases "Liebe" was used instead of "Sehr geehrte".
That particular case got apparently solved by adjusting the CMS wiki pages but I don't know is this the way it will/should be continued etc. So I am just leaving this as a random comment. Have no idea how this could be covered here.

**Comment by wmde-manicki** [comment]
Oh, and also there is this who-knows-when-coming-true idea of having donation website in English. That would I believe include translatable email contents.

File: src/Presenters/GreetingGenerator.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * @license GNU GPL v2+
+ * @author Gabriel Birke < gabriel.birke@wikimedia.de >
+ *
+ * @todo Make this translatable
+ */
+class GreetingGenerator {
+
+	public function createGreeting( string $lastName, string $salutation, string $title ): string {
+		if ( empty( $lastName ) ) {
```
Comment (JeroenDeDauw) [inline src/Presenters/GreetingGenerator.php]: #98

**Comment by gbirke** [comment]
@wmde-manicki For the formal/informal greeting texts we may create another class when needed.

The possible translatability and making it unit-testable of the greetings was the reason I put this in a class and not in a template. However, since this is so much presentation-related, it might move in the future. 

What's your take in this, @JeroenDeDauw, regular PHP class (with the need for unit tests and injection) or template?

**Comment by wmde-manicki** [comment]
> For the formal/informal greeting texts we may create another class when needed.

Sure.
What I wanted to to point out that given there are quite some different emails sent from the app (I didn't really count but I guess that might even be around 10? different emails, per language). Currently they share greetings part, as far as I remember, ie. they either use the same or simply say nothing more than "Howdy" (which might be explicitly included in the template).
But this case with some membership-related email needing different greeting shows that things might not always be that easy. But the generator still should be capable to handle other cases.

To sum up: I am not against the change you suggested. And good that you already have thought about possible "internationalization" of those things. Old codebase would probably require a series of nasty hacks to have it done.

**Comment by JeroenDeDauw** [comment]
> But this case with some membership-related email needing different greeting shows that things might not always be that easy. But the generator still should be capable to handle other cases.

I don't know what is or will be needed here, however two things come to mind when I read that:

* If there is an easy case and a difficult case, then keeping them apart, so that the easy case does not depend on the difficult one, might be good.
* Doing the simplest thing that works for the features currently being created and refactoring later if additional stuff is needed rarely hurts. (YAGNI basically)

**Comment by wmde-manicki** [comment]
> - If there is an easy case and a difficult case, then keeping them apart, so that the easy case does not depend on the difficult one, might be good.
> - Doing the simplest thing that works for the features currently being created and refactoring later if additional stuff is needed rarely hurts. (YAGNI basically)

I agree with both points, there is nothing to be changed in the proposed solution from my perspective. Having thought about all those greetings again I realized that there might not really be that much of hassle when some changes are required. English version(s) might not even need a class doing some magic, template would probably all cases.
So +1
